src/tensor_network.jl

```diff
@@ -429,4 +403,32 @@ function get_hyperedges(tn::TensorNetwork)::Array{Array{Symbol, 1}, 1}
         end
     end
     hyperedges
+end
+
+"""
+    find_connected_indices(tn::TensorNetwork, bond::Index)
+
+Given a tensor network and an index in the network, find all indices that are related via hyper edge
+relations. Involves recurisively checking bonds connected to neighbouring tensors of any newly
+related edges found. Returns an array in all edges in the group including the intial edge.
+"""
+function find_connected_indices(tn::TensorNetwork, bond::Index)
+    tensors_to_visit = Set{Symbol}()
+    push!.([tensors_to_visit], tn[bond])
+    related_edges = Set{Index}([bond])
+    while length(tensors_to_visit) > 0
+        tensor_sym = pop!(tensors_to_visit)
+        for g in hyperindices(tn[tensor_sym])
+            if length(intersect(related_edges, g)) > 0
+                new_edges = setdiff(g, related_edges)
+                for e in new_edges
+                    push!(related_edges, e)
+                    for t in tn[e]
+                        push!(tensors_to_visit, t)
+                    end
+                end
+            end
+        end
+    end
+    collect(related_edges)
 end
```

Comment:
I think the logic used in `find_connected_indices` is similar to some of the logic used in `get_hyperedges` so I'm wondering if the later should use the former to simplify it. Something like:

```
function get_hyperedges(tn)
    hyperedges = []
    edges = collect(bonds(tn))
    for edge in edges
        hyper_index = find_connected_indices(tn, edge)
        append!(hyperedges, unioin(map(x -> tn[x], hyper_index))
        setdiff!(edges, hyper_index)
    end
    hyperedges
end
```

Reply:
Yes it would be good to reuse the logic as would result in more code overall. Could be good to check that this will give equivalent results to the existing `get_hyperedges` method. I'll try adding this now and check a few cases.

For the above implementation wonder if one needs to be careful about modifying `edges` when also looping over it. Think this can lead to some undefined behaviour. Replacing with a while loop might make more sense.

Reply:
@brenjohn , I have implemented this function as you suggested and also added `hyperindices` and `tensor_data` functions that take into account the global hyperindices.